### PR TITLE
Call inflateEnd / deflateEnd in finalizers

### LINKED ISSUE #41: The library can leak
It's not entirely evident from the zlib docs but once you have called the `{inflate,deflate}Init` function you need to have a call to `{inflate,deflate}End` otherwise you will leak internally allocated structures. 

As such the finalizer [here](https://github.com/xavierleroy/camlzip/blob/3b0e0a5f774a86882bb45235140d550d2b5b481f/zlibstubs.c#L55-L63) should be adapted (perhaps easiest is to have a different finalizer for inflate and deflate initialized streams).

The following loops can be used to exhibit the behaviour: 

```ocaml
let rec loop_leak () = ignore (Zlib.inflate_init false); loop_leak ()

let rec loop_noleak () =
  let s = Zlib.inflate_init false in
  Zlib.inflate_end s; loop_noleak ()

let () = loop_leak ()
```

### PR DESCRIPTION
**Pull request by xavierleroy**
Otherwise, a memory leak occurs if `Zlib.deflate_end` / `Zlib.inflate_end` were not called before the `Zlib.stream` became unreachable.

Also: use `caml_alloc_custom_mem` instead of `caml_alloc_custom`.

Fixes: #41

### CODE REVIEW

**Comment by xavierleroy** [comment]
From a quick look at the implementation of `inflateEnd`, it seems harmless to call it twice, first in `Zlib.inflate_end` (where its return code will be checked) and then again in the finalizer (where it will return an error, which will be ignored).  Likewise for `deflateEnd`.

File: zlibstubs.c [hunk 1]
```diff
@@ -130,9 +126,28 @@ value camlzip_deflateEnd(value vzs)
...
-  value vzs = camlzip_new_stream();
+  value vzs =
+    caml_alloc_custom_mem(&camlzip_istream_ops,
+                          sizeof(z_streamp), sizeof(z_stream));
+  ZStream_val(vzs) = caml_stat_alloc(sizeof(z_stream));
+  ZStream_val(vzs)->zalloc = NULL;
+  ZStream_val(vzs)->zfree = NULL;
+  ZStream_val(vzs)->opaque = NULL;
+  ZStream_val(vzs)->next_in = NULL;
+  ZStream_val(vzs)->next_out = NULL;
```
Comment (dbuenzli) [inline zlibstubs.c]: Note the [manual](https://www.zlib.net/manual.html) about of `inflateInit2` says:

>  The fields next_in, avail_in, zalloc, zfree and opaque must be initialized before by the caller.

So you can likely remove the `next_out` and add `ZStream_val(vzs)->avail_in = 0`.

(For `deflateInit2` above the manual only mentions `zalloc`, `zfree` and `opaque` but it seems harmless to zero more).

**Review by dbuenzli — APPROVED** [review]
Except for a little nit. Looks good to me.

File: zlibstubs.c (hunk 1 above)
Comment (xavierleroy) [reply zlibstubs.c]: OK, fixed in 237dd2d.

**Pull request merged**